core: Atomic counters

gebart opened the pull request:
This PR adds `atomic_inc` and `atomic_dec` as discussed in #2302 

I have created basic implementations for some of the CPUs as an example.

kaspar030 commented on cpu/msp430-common/atomic.c:
```diff
@@ -30,3 +31,21 @@ unsigned int atomic_set_return(unsigned int *val, unsigned int set)
+int atomic_inc(int *val)
+{
+    int old_val;
+    dINT();
```
this should be a pair of state=disableIRQ()/restoreIRQ(state)

gebart replied:
I did not check if the msp430 platform was up to date with regards to IRQ functions, since the above atomic_set_return also used eINT/dINT.

I will change this to match the Cortex-M0 implementation.

kaspar030 commented:
The 'naive' implementation looks like a good candidate for a generic fallback implementation that could be used whenever an architecture doesn't define an optimized version.

Could you do something  like, ```#ifndef ARCH_HAS_ATOMIC_INT; #define atomic_inc atomic_inc_generic; #endif```?

gebart commented:
@kaspar030 I get the feeling that your suggested `#ifdef` will only make the platform configuration more verbose. Do we really have that many CPU architectures that we need this? Otherwise I suggest to just copy the generic implementation to each CPU family as a starting point, then write optimized versions for each CPU whenever someone feel like improving it.

kaspar030 commented:
@gebart Well, I prefer one ```#define ARCH_HAS_ATOMIC_INC``` in the per-cpu header than a copied generic version in the generic case. Seems more natural: initially, we don't have a specialized version. Without any change, the generic version will be used. Once someone 'overrides' the generic version, he adds the corresponding define.

That way, if any problem shows up, it's also easy to disable any specialized function by commenting out the define...

gebart commented:
@kaspar030 I updated the PR according to your suggestions. I also removed copies of the generic `atomic_arch_set_return` to reduce code duplication further. The generic interface is defined in `core/atomic.c` and protected by an `#if ARCH_HAS_ATOMIC_INC != 0`

Missing is an implementation for x86 (`x86_atomic.c`) because I am not familiar enough with x86 assembly language to implement it.

gebart commented:
should I remove the API CHANGE label? This PR only adds functions to the core API.

haukepetersen commented:
I just stumbled upon a use case, where you need to increment/decrement a value atomically by a given number (e.g. inc by 3). What is your opinion for including this into the proposed API, say something like
```c
int atomic_arch_inc(int *val, int howmuch);
```
?

gebart commented:
@haukepetersen What is your use case?

Maybe we should introduce @Kijewski 's suggestion of Compare And Swap?

Then you would write:

```c
atomic_cas(&val, val, val+3);
```

My first use case is: Atomic counter as a counting semaphore to introduce inhibiting of low power modes when modules are using hardware devices which need the CPU to stay alive. i.e. a reference count on power modes. In this case we only need up/down counting in steps of 1.

haukepetersen commented:
My use case is a simple usage counter, that keeps track of the number of threads that are being passed a (read-)pointer to a data structure. If a thread is done reading from it, it decrements the counter again (atomically, of course). If I send this pointer to 4 different threads, I have to increment it (atomically) 4 times before sending. This would be best looking with such function...

I like @Kijewski 's suggestion, works very well for me. Does it even make sense to just have this one function, as you can do everything we need with it?

gebart commented:
@haukepetersen , @Kijewski , it does make sense to implement `atomic_cas`.
Any objections to rewriting this code as @Kijewski 's suggestion in https://github.com/RIOT-OS/RIOT/issues/2302#issuecomment-70238314 ?

I would be leaving `atomic_cas` as a public function which can be used for special cases of atomic variables, such as @haukepetersen 's use case, but the simplest way to use the API for everyday use would still be to call either `atomic_inc`/`atomic_dec` or `atomic_set`/`atomic_reset`.

haukepetersen commented:
+1 for implementing `atomic_cas`.

But I would really think of slimming the interface down to this one function, for the sake of resource efficiency. This is a kernel function that is normally only used by people who know what they are doing, it's not an everyday interface as e.g. threading or mutexes...

kaspar030 commented:
@haukepetersen Well, if you need more than one line to do atomic_inc, something's wrong with the API.
And you need that if there's only atomic_cas, right?

When we're touching this, we should think about introducing special types for atomic values.

gebart commented:
@haukepetersen I liked the `static inline` version of the counters that @Kijewski wrote. Makes the code using it clean and simple to understand, without knowing the details of atomic_cas and why it has to be inside a loop

haukepetersen commented:
ok, I guess you convinced me...

gebart commented:
@kaspar030 how do you propose these atomic types would be used? 
If I wrote this in C++ I would overload a bunch of operators on them to keep the code clear and readable, but I guess in C we'll have to do something similar to byteorder.h, or am I missing something?

gebart commented:
simply doing the following will not warn if passed a regular `int` variable:

```c
typedef int atomic_int_t;
```

How can I enforce strict type checking in C? Should I change atomic_int_t into a union instead?

gebart commented:
Rebased and updated with a generic atomic_cas function for all atomic operations as per @Kijewski 's suggestion.

TODO (in a separate PR): Update mutex.c etc to use atomic_int_t and atomic_set/clear instead of old atomic_set_return.

Kijewski commented on core/atomic.c:
```diff
@@ -0,0 +1,53 @@
+int atomic_cas(atomic_int_t *var, int old, int now)
+{
+    unsigned int mask = disableIRQ();
+    if (var->val != old) {
```
you need to restore the IRQs.

gebart replied:
Thank you, updated the PR accordingly.

Kijewski commented:
For x86 atomic_cas is:
```c
int atomic_cas(atomic_int_t *dest, int known_value, int new_value)
{
    uint8_t successful;
    asm volatile ("lock cmpxchgl %2, %0\n"
                  "seteb %1"
                  : "=m"(dest->value), "=g"(successful)
                  : "r"(new_value), "a"(known_value)
                  : "flags");
    return successful;
}
```

gebart commented:
@Kijewski I added the atomic_cas that you posted here and updated the copyright on x86_atomic.c in your name

Kijewski commented:
@gebart , thanks!

gebart commented:
I fixed a typo/bug and made some small cleanups of the M3/M4 implementations.

Ready for review IMO.

gebart commented:
ping @kaspar030

Kijewski commented on core/include/atomic.h:
```diff
@@ -35,9 +46,91 @@
+ *
+ * @return The value of *val* before the increment.
+ */
+static inline int atomic_inc(atomic_int_t *var) {
```
Newline before `{` here at at the other instances.

gebart commented:
Updated with `astyle` and rebased on master

authmillenon commented:
Anyone care to review?

kaspar030 commented on core/include/atomic.h:
```diff
@@ -35,9 +46,101 @@
+ * @return 1 if the old value was 0 and the variable was successfully updated
+ * @return 0 if the variable was already set
+ */
+static inline int atomic_set(atomic_int_t *var)
```
Hm, I find the name misleading. I started commenting the brief description until I realized that you actually meant "set the variable to 1".
Also, where is the difference to ```atomic_set_return(var,1)```? Isn't this a convenience function to save that variable?

OlegHahm replied:
@gebart, I would agree with @kaspar030 here.

kaspar030 commented on core/include/atomic.h:
```diff
@@ -35,9 +46,101 @@
+ * @return 0 if the variable was already cleared
+ */
+static inline int atomic_clear(atomic_int_t *var)
+{
```
see above.

kaspar030 commented on core/include/atomic.h:
```diff
@@ -35,9 +46,101 @@
+/**
+ * @brief Atomic Compare and Swap
+ *
+ * Updates the value in var iff the current value of var is equal to old.
```
iff -> if

gebart replied:
on purpose. "iff" = "if and only if"

kaspar030 commented on core/include/atomic.h:
```diff
@@ -35,9 +46,101 @@
  */
 unsigned int atomic_set_return(unsigned int *val, unsigned int set);
```
I just checked, looks like there are only two or three uses of ```atomic_set_return```. We should (maybe as a follow-up PR) rewrite this to use ```atomic_int_t``` and update the places where it is used.

gebart commented:
My thought was to deprecate `atomic_set_return` in the long run and use `atomic_cas` instead.
The current uses of `atomic_set_return` are found in `core/mutex.c` and `sys/posix/pthread/pthread_spin.c` where it is being used as the `atomic_set`/`atomic_clear` pair to handle a semaphore variable atomically.

`tests/unittests/tests-core/tests-core-atomic.c` should be updated as well to test the `atomic_int_t` based functions as well.

gebart commented:
@kaspar030 ping

I added unit tests for the atomic counter functions and for `atomic_cas`.

Removing atomic_set_return should be done as a follow up PR.

Ready for review.

gebart commented:
- Added ATOMIC_INIT initializer macro, similar to MUTEX_INIT.
- Added ATOMIC_VALUE getter macro to access the internal integer variable directly.

gebart commented:
I'd like to get this merged so that we can proceed to add LPM support for Kinetis in #2605. Since this is core it is necessary that at least two maintainers ack this before merge. Does anyone have the time to review this?

kaspar030 commented:
ACK from my side. Sorry for not getting back to this earlier.
@Kijewski Maybe you can take a second look?

gebart commented:
Can I squash this to a somewhat more manageable number of commits?

kaspar030 commented:
yeah. I'll do some more intense testing tomorrow and the find the second ack.

gebart commented:
squashed, modified implementations to use `ATOMIC_VALUE` to access the internal variable inside the atomic_int_t. The follow up PR will additionally use this macro to check the value of mutexes

OlegHahm commented:
There are still some squash commits.

gebart commented:
the unsquashed commits are the changes I made after @kaspar030 's ACK

OlegHahm commented:
Ah, ok, sorry.

kaspar030 commented:
The files cortex-m3_common/atomic_arch.c and cortex-m4_common/atomic_arch.c are the same.

kaspar030 commented:
But so are many others in cortex-m3_common/ and cortex-m4_common/. Let's unify in a different PR.

kaspar030 commented:
Re-ACK after squash.

gebart commented:
This is a core change, can we get a second ACK? @haukepetersen @OlegHahm @Kijewski ?

gebart commented:
squashed

OlegHahm commented:
What about arm7? It implements `atomic_set_return` but not `atomic_cas`. (Or did I overlook something?)

gebart commented:
@OlegHahm you are correct, however I do not know enough of arm7 assembly to implement it in atomic.s. Does anyone have a simple implementation of compare and swap for ARM7 ready?

kaspar030 commented:
@gebart you mean ARMv7?

gebart commented:
@kaspar030 no, ARMv4

OlegHahm commented:
@gebart, can we not just use the software (C) implementation for compare-and-swap on ARM7, but still use the optimized set_return? As far as I can tell, there's no way to improve compare-and-swap without corresponding instructions in the CPU.

Kijewski commented:
To have overridable default functions, you should use weak aliases: https://github.com/Kijewski/RIOT/commit/d5f753aa58e023a5fa3aa5f557a5ac5ed4cc3541

OlegHahm commented:
We're already using weak aliases at several points, but we might want to get rid of it. After all we're claiming RIOT to be ANSI C99 compliant.

Kijewski commented:
>  but we **might** want to get rid of it

Emphasis is mine. We **might** want to embrace pragmatism over dogmatism. :)

OlegHahm commented:
Agreed, I'm just wondering (thinking aloud) if we're sure that no one wants to build RIOT with a different compiler than GCC or clang.

Kijewski commented:
@OlegHahm, please see https://github.com/Kijewski/RIOT/commit/9b333af85fc8c0030846c3913ed2758b04f989ab. Do you think this is a better option?
The Visual Studio variant should work but is (of course) untested.

OlegHahm commented:
Actually, I'm not sure. Keil, IAR, and Crossworks seem to support `#pragma weak`. However, I think what's written on Wikipedia is (in this case) true:
> Weak symbols are not mentioned by C or C++ language standards; as such, inserting them into code is not very portable. Even if two platforms support the same or similar syntax for marking symbols as weak, the semantics may differ in subtle points, e.g. whether weak symbols during dynamic linking at runtime lose their semantics or not.

Hence, even if we add an `#ifdef` for every compiler we might end up in a situation where code behaves differently on different platforms. Since this is core, I would be very careful here.

OlegHahm commented:
Sounds good to me.

gebart commented:
Updated to have two separate ifdefs.

OlegHahm commented:
Can't we use basically the same assembler for the Cortex platforms? Otherwise, we should remove the set-and-return code there since it is same as in the platform independent solution.

gebart commented:
@OlegHahm Good idea. Updated.

I have a follow up PR for removing atomic_set_return from the current users (mutex, pthread, ng_pktbuf) waiting in the pipe.

OlegHahm commented:
@gebart, are you planning to rename to `atomic_set()` to `atomic_set_to_one()`? I like short names and coming from the hardware perspective, it may be natural to think of a set function this way, but I think for other people the current name may indeed be misleading.

gebart commented:
On May 6, 2015 9:29 AM, "Oleg Hahm" <notifications@github.com> wrote:
>
> @gebart, are you planning to rename to atomic_set() to
atomic_set_to_one()? I like short names and coming from the hardware
perspective, it may be natural to think of a set function this way, but I
think for other people the current name may indeed be misleading.

I see your point. I'll update the PR when I get back to my desktop.

>
> —
> Reply to this email directly or view it on GitHub.

gebart commented:
Updated with renamed functions. Also deleted the native implementation of atomic_set_return because it was the same as the generic implementation except for a (disabled) debug print.

Kijewski commented on cpu/arm7_common/include/arm_cpu.h:
```diff
@@ -6,8 +6,8 @@
-#ifndef _ARM_CPU_H
-#define _ARM_CPU_H
+#ifndef ARM_CPU_H_
+#define ARM_CPU_H_
```
`%s,#endif // _ARM_CPU_H,#endif /* ARM_CPU_H_ */,`

gebart replied:
done.

Kijewski commented:
But for my last comment the PR should be ready to merge.

gebart commented:
@kaspar030 ACK still holds?

OK to squash?

kaspar030 commented:
Yes!

gebart commented:
Squashed and rebased, waiting for Travis.
@Kijewski Can I take your comment as an ACK?

OlegHahm commented:
Travis fell asleep for arm7 build tests. Kicked it.

OlegHahm commented:
ACK and go.
